src/zenml/integrations/vertex/orchestrators/vertex_orchestrator.py

```diff
@@ -0,0 +1,443 @@
+    from zenml.runtime_configuration import RuntimeConfiguration
+    from zenml.stack import Stack
+    from zenml.steps import BaseStep
+
+
+logger = get_logger(__name__)
+
+
+def _clean_pipeline_name(pipeline_name: str) -> str:
+    """Clean pipeline name to be a valid Vertex AI Pipeline name.
+
+    Arguments:
+        pipeline_name: pipeline name to be cleaned.
+
+    Returns:
+        Cleaned pipeline name.
+    """
+    return pipeline_name.replace("_", "-").lower()
+
+
+class VertexOrchestrator(BaseOrchestrator, GoogleCredentialsMixin):
+    """Orchestrator responsible for running pipelines on Vertex AI.
+
+    Attributes:
+        custom_docker_base_image_name: Name of the Docker image that should be
+            used as the base for the image that will be used to execute each of
+            the steps. If no custom base image is given, a basic image of the active
+            ZenML version will be used. **Note**: This image needs to have ZenML
+            installed, otherwise the pipeline execution will fail. For that reason,
+            might want to extend the ZenML Docker images found here:
+            https://hub.docker.com/r/zenmldocker/zenml/
+        project: GCP project name. If `None`, the project will be inferred from
+            the environment.
+        location: Name of GCP region where the pipeline job will be executed.
+            Vertex AI Pipelines is available in the following regions:
+            https://cloud.google.com/vertex-ai/docs/general/locations#feature-availability
+        pipeline_root: a Google Cloud Storage path that will be used by the Vertex
+            AI Pipelines.
+        encryption_spec_key_name: The Cloud KMS resource identifier of the customer
+            managed encryption key used to protect the job. Has the form:
+            `projects/my-project/locations/my-region/keyRings/my-kr/cryptoKeys/my-key`.
+            The key needs to be in the same region as where the compute resource
+            is created.
+        workload_service_account: the service account for workload run-as account.
+            Users submitting jobs must have act-as permission on this run-as account.
+            If not provided, the default service account will be used.
+        network: the full name of the Compute Engine Network to which the job should
+            be peered. For example, `projects/12345/global/networks/myVPC`
+            If not provided, the job will not be peered with any network.
+        service_account_path: path to service account file that will be used to create the
+            Vertex AI Pipelines job. If not provided, falls back to default credentials.
+        synchronous: If `True`, running a pipeline using this orchestrator will
+            block until all steps finished running on Vertex AI Pipelines service.
+    """
+
+    custom_docker_base_image_name: Optional[str] = None
+    project: Optional[str] = None
+    location: str
+    pipeline_root: str
```

Comment:
After taking a second look at the attributes I have another question here: This `pipeline_root` attribute specifies a GCP bucket where vertex will write stuff like their logs etc right? If that's the case, do you think it makes sense to make this optional and fallback to a subdirectory in the zenml artifact store (in case it's a gcp bucket, which we could validate)?

Reply:
The `pipeline_root` is a Cloud Storage URI where Vertex will save the artifacts generated by the steps. The outputs can be specified using the [component specification](https://www.kubeflow.org/docs/components/pipelines/sdk-v2/component-development/#define-your-components-interface) or the [Python function-based components](https://www.kubeflow.org/docs/components/pipelines/sdk-v2/python-function-components/), but we are not relying on Vertex to handle the artifacts.

It would be good if `pipeline_root` is `Optional`, but the only case that we can fall back to a subdirectory in the zenml artifact store is when the artifact store used is a `GCPArtifactStore`, right? Maybe we can leave this `Optional`:

1. If the user has not provided a value for `pipeline_root` but is using a `GCPArtifactStore` as an artifact store, then fall back to a subdirectory in the path of the artifact store.
2. If the user has not provided a value for `pipeline_root` and he is not using `GCPArtifactStore` as an artifact store, then raise an error indicating to him that he should specify a value for `pipeline_root` if he is not using `GCPArtifactStore`.

Reply:
Yes exactly, so the stack validation would only work if
- `pipeline_root` is set or
- the artifact store of the stack is a `GCPArtifactStore` (which I guess it would be in the majority of cases where people run on vertex)

Reply:
Done! I needed to add an auxiliary "private" attribute called `_pipeline_root` because I could not modify `pipeline_root` as the `StackComponent` class does not allow to mutate the value of its attributes.